### PR TITLE
Fix precedence for PATH for build-tools-depends

### LINKED ISSUE #8951: Installed executables take precedence over `build-tool-depends`
**Describe the bug**
Executables from `~/.local/bin`[^1] shadow executables that are specified via `build-tool-depends`.

Why is this a problem? How can this corrupt your builds in non-obvious ways? How much effort is it to debug this? I'm not going into rant mode now, please use your own judgment.

Instead, I will describe the underlying reasons for this bug and possible solutions in a follow-up comment.

**To Reproduce**
A minimal skeleton that can be used to reproduce the issue is at https://github.com/sol/cabal-bug/tree/build-tools

You may need to adapt this according to your OS and whether you have an existing `~/.cabal` or not. So make sure that you fully understand what it is doing before you run it.

**System information**
 Operating system: Linux
cabal: 3.10.1.0
ghc: 9.6.1

[^1]: Note that this may be `~/.cabal/bin` on your system.

### PR DESCRIPTION
**Pull request by gbaz**
Resolves #8951 and modifies #8506 in the process. This is very subtle stuff and definitely needs a regression test added as well.

* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

### CODE REVIEW

**Comment by andreabedini** [comment]
Well done troubleshooting this and coming up with a fix! 

In the issue, you say (emphasis mine):

> the code is very gnarly around how this stuff works, so _need to think carefully_ on best way to fix it.

and right after:

> My _proposed, janky-as-heck, solution_ is to have the path from extra-prog-path eagerly added at the end of the useExtraPathEnv option in SetupScriptOptions.

If you allow me, this gives me the impression that your solution somewhat falls short of your own expectations and it makes me  wonder if togheter we can find another way.

From your diagnosis:

> The deeper diagnosis is that invoke in SetupWrapper.hs is called by v2-build to run the actual build, performed by Cabal-the-library. It sets a PATH including the paths of tools created in build-tool-depends at the head. Then the Setup in Cabal-the-library in turn adds the extra-prog-path stuff even further, in front of that.

Repeating in my own words: cabal-install puts extra-prog-path in the PATH, then passes extra-prog-path to Cabal, which also puts it in the PATH causing confusion.

What if cabal-install did not pass extra-prog-path to Cabal at all? using it only to set the PATH before calling Cabal.
Those who call Cabal directly could use that option on their own (or set PATH as they like). Would this be a bad solution?

> This means that the usefulness of 8506 is lost in the non v2-build case

https://github.com/haskell/cabal/pull/8506 was to resolve https://github.com/haskell/cabal/issues/6304, which was raised as a v2-build problem. Was there even a v1-build issue to solve?

What is the difference between 
```
./Setup.hs configure --extra-prog-path /x/y/z ...
```
and
```
PATH=/x/y/z:$PATH ./Setup.hs configure ...
```

?


I am sure I am missing something here; so I thank you in advance for helping me to understand.

**Comment by Kleidukos** [comment]
@gbaz Would you like my help with fixing the conflicts of your PR?

**Review by Kleidukos — APPROVED** [review]

File: cabal-install/src/Distribution/Client/ProjectPlanning.hs [hunk 1]
```diff
@@ -4009,8 +4012,11 @@ setupHsScriptOptions
       , useDistPref = builddir
       , useLoggingHandle = Nothing -- this gets set later
       , useWorkingDir = Just srcdir
-      , useExtraPathEnv = elabExeDependencyPaths elab
-      , useExtraEnvOverrides = dataDirsEnvironmentForPlan distdir plan
+      , useExtraPathEnv = elabExeDependencyPaths elab ++ elabProgramPathExtra
+      , -- note that the above adds the extra-prog-path directly following the elaborated
+        -- dep paths, so that it overrides the normal path, but _not_ the elaborated extensions
+        -- for build-tools-depends.
+        useExtraEnvOverrides = dataDirsEnvironmentForPlan distdir plan
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/ProjectPlanning.hs:4019]: Comma goes here (for a cleaner diff), I believe.

File: Cabal/src/Distribution/Simple/Configure.hs [hunk 2]
```diff
@@ -1028,10 +1028,17 @@ mkProgramDb cfg initialProgramDb = programDb
...
-      map
-        ProgramSearchPathDir
-        (fromNubList $ configProgramPathExtra cfg)
-        ++ getProgramSearchPath initialProgramDb
+      getProgramSearchPath initialProgramDb
+        ++ map
+          ProgramSearchPathDir
+          (fromNubList $ configProgramPathExtra cfg)
+
+-- Note. We try as much as possible to _prepend_ rather than postpend the extra-prog-path
```
Comment (ffaf1) [inline Cabal/src/Distribution/Simple/Configure.hs:1036]: maybe postpend → append

**Review by sol — CHANGES_REQUESTED** [review]
From what I understand, this PR includes unrelated formatting changes, making a review more work than it would need to be.

**Comment by ffaf1** [comment]
@sol I believe a rebase will fix those.

File: Cabal/src/Distribution/Simple/Configure.hs (hunk 2 above)
Comment (gbaz) [reply Cabal/src/Distribution/Simple/Configure.hs:1036]: append doesn't specify to front or back

Comment (ffaf1) [reply Cabal/src/Distribution/Simple/Configure.hs:1036]: I am not going to debate a native speaker, but “add (something) to the end of a written document.” ([def](https://dictionary.cambridge.org/dictionary/english/append)).

Comment (ulysses4ever) [reply Cabal/src/Distribution/Simple/Configure.hs:1036]: Fun! I also don't like how artificial "postpend" sounds but on the other hand maybe extra clarity is justified here...

Comment (geekosaur) [reply Cabal/src/Distribution/Simple/Configure.hs:1036]: FWIW I was surprised by "append doesn't specify"; I've always understood it as meaning "to the end of".

File: cabal-install/src/Distribution/Client/ProjectPlanning.hs (hunk 1 above)
Comment (sol) [reply cabal-install/src/Distribution/Client/ProjectPlanning.hs:4019]: Yes, please.

**Comment by ulysses4ever** [comment]
@gbaz do you think you give some more love to this effort? It looks close!

**Comment by andreasabel** [comment]
An end-to-end test that could have verified (or likely falsified) this PR can be harvested from:
- #9756

**Pull request merged**